File: inzynierka_backend/src/UserService.py
from src.IUserRepository import IUserRepository


class UserService:
    def __init__(self, repository: IUserRepository):
        self._repository = repository

    def get_user(self, email: str):
        return self._repository.get_user(email)

    def change_password(self, email: str, new_password_hash: str):
        user = self.get_user(email)
        return self._repository.modify_account(user.id, password_hash=new_password_hash)

    def change_email(self, old_email: str, new_email: str):
        user = self.get_user(old_email)
        return self._repository.modify_account(user.id, email=new_email)

    def get_users(self):
        return self._repository.get_users()

    def promote_user_to_admin(self, email):
        return self._repository.promote_user_to_admin(email)

    def verify_user(self, email):
        return self._repository.verify_account(email)

    def create_user(self, email: str, password_hash: str, disabled: bool, is_admin: bool):
        return self._repository.create_user(email, password_hash, disabled, is_admin)

File: front_inzynierka/src/main/kotlin/com/inzynierka/data/service/UserService.kt
package com.inzynierka.data.service

import com.inzynierka.common.DomainError
import com.inzynierka.common.Result
import com.inzynierka.data.parsedRemoteExceptionMessage
import com.inzynierka.data.repository.IUserRepository
import com.inzynierka.domain.core.UserData
import com.inzynierka.domain.service.IUserService

class UserService(private val userRepository: IUserRepository) : IUserService {

    override suspend fun loginUser(email: String, password: String): Result<Unit> {
        return try {
            Result.Success(userRepository.loginUser(email, password))
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun promoteUserToAdmin(email: String): Result<Unit> {
        return try {
            Result.Success(userRepository.promoteUserToAdmin(email))
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun registerUser(email: String, password: String): Result<Unit> {
        return try {
            Result.Success(userRepository.registerUser(email, password))
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun verifyAccount(code: String): Result<Unit> {
        return try {
            Result.Success(userRepository.verifyAccount(code))
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun resendVerificationCode(): Result<Unit> {
        return try {
            Result.Success(userRepository.resendVerificationCode())
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun getUserData(): Result<UserData> {
        return try {
            val data = userRepository.getUserData()
            Result.Success(
                UserData(disabled = data.disabled, isUserAdmin = data.isAdmin)
            )
        } catch (e: Throwable) {
            Result.Error(
                DomainError(e.parsedRemoteExceptionMessage)
            )
        }
    }

    override suspend fun changePassword(newPassword: String, oldPassword: String): Result<Unit> {
        return try {
            Result.Success(userRepository.changePassword(newPassword, oldPassword))
        } catch (e: Throwable) {
            Result.Error(DomainError(e.parsedRemoteExceptionMessage))
        }
    }

    override suspend fun changeEmail(email: String): Result<Unit> {
        return try {
            Result.Success(userRepository.changeEmail(email))
        } catch (e: Throwable) {
            Result.Error(DomainError(e.parsedRemoteExceptionMessage))
        }
    }

    override fun logout() {
        userRepository
    }

}